tests/unit/testsuite/Bolt/Boltpay/Block/Checkout/BoltpayTest.php

```diff
@@ -255,6 +286,91 @@ public function testGetSaveOrderURL()
         $this->assertEquals($expected, $result);
     }
 
+    /**
+     * @test
+     * that a proper product key is returned for every config or an exception is thrown
+     *
+     * @covers ::getPublishableKeyForThisPage
+     * @dataProvider getPublishableKeyForThisPageProvider
+     *
+     * @param string|null $multiStepKey             the multi-step publishable key, empty string or null indicates not set
+     * @param string|null $paymentOnlyKey           the payment-only publishable key, empty string or null indicates not set
+     * @param string      $routeName                The magento path route name
+     * @param string      $controllerName           The magento path controller name
+     * @param string|null $predictedReturnValue     The key that is expected to be returned by the mock or null if not returned
+     *
+     * @throws Exception if there is a failure retrieving the focus mock's Request object
+     */
+    public function getPublishableKeyForThisPage($multiStepKey, $paymentOnlyKey, $routeName, $controllerName, $predictedReturnValue )
+    {
+        $this->currentMock = $this->getMockBuilder('Bolt_Boltpay_Block_Checkout_Boltpay')
+            ->setMethods(array('getPublishableKey'))
+            ->disableOriginalConstructor()
+            ->disableOriginalClone()
+            ->disableArgumentCloning()
+            ->getMock()
+        ;
+        $this->currentMock->getRequest()
+            ->setRouteName($routeName)
+            ->setControllerName($controllerName);
+
+        $mockProxy = new Bolt_Boltpay_Block_Checkout_Boltpay();
+
+        $this->app->getStore()->setConfig('payment/boltpay/publishable_key_multipage', $multiStepKey);
+        $this->app->getStore()->setConfig('payment/boltpay/publishable_key_onepage', $paymentOnlyKey);
+
+        switch ($controllerName) {
+            case 'cart':
+                $routeName = 'catalog';
+                #intentionally falling through because this his handled the same way as catalog/product
+            case 'product':
+                if ($routeName === 'catalog') {
+                    $expectedReturnedValue = $multiStepKey ?: $paymentOnlyKey;
+                    break;
+                }
+                # intentionally falling through because the correct route was not found
+            default:
+                $expectedReturnedValue = $paymentOnlyKey ?: $multiStepKey;
+
+        }
+
+        $this->currentMock->expects($this->exactly(2))->method('getPublishableKey')
+            ->withConsecutive(
+                [$this->equalTo('multi-page')],
+                [$this->equalTo('one-page')]
+            )
+            ->willReturnOnConsecutiveCalls(
+                $mockProxy->getPublishableKey('multi-page'),
+                $mockProxy->getPublishableKey('one-page')
+            )
+        ;
+
+        if ($predictedReturnValue) {
```

Comment:
@mccottry My guess here is that you did this so that you could DRY the function testing here, and it's written out because you can't really test exception and non-exception code with a dataprovider's output. I would much rather have two unit test functions here instead of one for clarity though - one for an exception flow and one for a non-exception. @oldDriver is right in that we should try to avoid logic of this kind in the test.

Reply:
I agree with this assessment and the implementation was bothering me as I was thinking about a unit test for a unit test, which absolutely makes no sense.

@matt-thomason You guess is correct and preference is what will now be implemented.